Add license sniff and fixes

xtreamwayz opened the pull request:
The test are executed independent of the standard `AbstractSniffUnitTest` class. That class is only available if `PHP_CodeSniffer` is downloaded by composer with `prefer-source`.

Copyright year detection:
- [x] Detect current copyright year from LICENSE.md and/or COPYRIGHT.md
- [x] Detect current copyright year from all source files
- [x] Globally exclude */TestAssets/*

Violation detection (when running phpcs):
- [x] Check file-level docblock
- [x] Check missing/invalid see tag
- [x] Check missing/invalid copyright tag
- [x] Check missing/invalid license tag
- [x] Check order of see, copyright and license tags
- [x] Check for current copyright year in LICENSE.md and COPYRIGHT.md
- [x] Check text in LICENSE.md and COPYRIGHT.md to make sure it's consistence in all components

Fixes (when running phpcbf):
- [x] Create missing LICENSE.md and/or COPYRIGHT.md
- [x] Use the detected copyright year in LICENSE.md and COPYRIGHT.md
- [x] Fix invalid see tag
- [x] Fix invalid copyright tag
- [x] Fix invalid license tag
- [x] Update copyright in LICENSE.md and COPYRIGHT.md to the current year
- [x] Update the entire text in LICENSE.md and COPYRIGHT.md for consistency
- [x] Fix missing newline after comment

TODO:
- [ ] Fix tests
- [ ] Test on zendframework components
  - [ ] zend-expressive
  - [ ] zend-code
  - [ ] zend-expressive-tooling
  - [ ] zend-mvc
  - [ ] zend-http

Discussion:
- [x] Is the text for [COPYRIGHT.md](https://github.com/xtreamwayz/zend-coding-standard/blob/feature/license-sniff/src/ZendCodingStandard/Utils/LicenseUtils.php#L19-L24) correct?
- [x] Is the text for [LICENSE.md](https://github.com/xtreamwayz/zend-coding-standard/blob/feature/license-sniff/src/ZendCodingStandard/Utils/LicenseUtils.php#L26-L56) correct?
- [ ] How to include the ruleset in components?
  1. `<rule ref="vendor/zendframework/zend-coding-standard/ruleset.xml"/>`
  2. `<rule ref="vendor/zendframework/zend-coding-standard/src/ZendCodingStandard/ruleset.xml"/>`
  The first is like a wrapper and loads the second one anyway.

weierophinney commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+
+matrix:
+  include:
+    - php: 7
```
Maybe should also test it on 7.1? That way we'll know if there's breakage if/when we want to use it against that version instead of 7.0.

weierophinney commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+  - if [[ $TEST_COVERAGE != 'true' ]]; then phpenv config-rm xdebug.ini || return 0 ; fi
+
+install:
+  - travis_retry composer install $COMPOSER_ARGS --ignore-platform-reqs
```
Why `--ignore-platform-reqs` here? Is there a compelling reason?

xtreamwayz replied:
I've added the lowest/locked/latest strategy.

weierophinney commented on ruleset.xml:
```diff
@@ -1,23 +1,5 @@
-            <property name="ignoreBlankLines" value="false"/>
-        </properties>
-    </rule>
+    <rule ref="vendor/zendframework/zend-coding-standard/ZendCodingStandard/ruleset.xml"/>
```
A little confused by the changes here.

- In components that already use this package, this particular file is the one referenced. As an example, from zendframework/zend-expressive:

  ```xml
  <rule ref="./vendor/zendframework/zend-coding-standard/ruleset.xml"/>
  ```

  Shouldn't this then be the full ruleset, and not reference another ruleset?

- The referenced path would be potentially valid for components referencing the ruleset installed via Composer, _assuming that phpcs follows the reference_, but is malformed; `ZendCodingStandard` is a subdirectory of `src/`~

- The referenced path is invalid for _this_ repository (should be `src/ZendCodingStandard/ruleset.xml`), though I'm not sure if that's a problem or not.

I suspect the testing against other components will tell us for sure if this works as expected, but thought I'd highlight the potential issues first.

xtreamwayz replied:
> In components that already use this package, this particular file is the one referenced. As an example, from zendframework/zend-expressive... Shouldn't this then be the full ruleset, and not reference another ruleset?

As you say, in expressive, this file is imported: `vendor/zendframework/zend-coding-standard/ruleset.xml`.
However for PHP_CodeSniffer to work, a ruleset.xml file must be `some/path/<ruleset_name>/ruleset.xml`. I've tried several options and even including this in `src` directly. This is the only way PHP_CodeSniffer detects the ruleset as a valid ruleset. I also tried this with PHP_CodeSniffer 3.0RC with the same result.

I guess using twice the name ruleset.xml is confusing. The first (this) one includes the real standard ruleset. I guess including './vendor/zendframework/zend-coding-standard/src/ZendCodingStandard/ruleset.xml' might also work, but than all components need another PR.

> The referenced path is invalid for this repository (should be src/ZendCodingStandard/ruleset.xml), though I'm not sure if that's a problem or not.

Since this ruleset.xml file is only meant for inclusion in other projects, it doesn't matter. Running phpcs inside this project will use phpcs.xml.

I have tested this against another [repo](https://github.com/xtreamwayz/zend-coding-test) and it worked as it is. I'll test zf components later on.

xtreamwayz replied:
@weierophinney There was actually an error in that file. the correct location would be `<rule ref="vendor/zendframework/zend-coding-standard/src/ZendCodingStandard/ruleset.xml"/>`.

weierophinney commented on src/ZendCodingStandard/Sniffs/Commenting/FileLevelDocBlockSniff.php:
```diff
@@ -0,0 +1,295 @@
+
+    private $copyrightChanged;
+
+    private $fix;
```
The above should have `@var` annotations declaring types.

weierophinney commented on src/ZendCodingStandard/Sniffs/Commenting/FileLevelDocBlockSniff.php:
```diff
@@ -0,0 +1,295 @@
+            true
+        );
+
+        $ignore = [
```
Maybe this declaration should be as a constant or class property?

weierophinney commented on src/ZendCodingStandard/Sniffs/Commenting/FileLevelDocBlockSniff.php:
```diff
@@ -0,0 +1,295 @@
+                LicenseUtils::getLicenseFile(),
+                $this->licenseFirstYear,
+                $this->licenseLastYear
+            );
```
Maybe extract the above into its own method?

weierophinney commented on src/ZendCodingStandard/Sniffs/Commenting/FileLevelDocBlockSniff.php:
```diff
@@ -0,0 +1,295 @@
+     *                  called again on the current file until the returned stack
+     *                  pointer is reached. Return (count($tokens) + 1) to skip
+     *                  the rest of the file.
+     */
```
Nitpick: I've been _not_ aligning types/names in annotations, and indenting any line continuations only by four characters. That allows more text per continuation line, and is easier to format overall (changes in types and/or variable name lengths then do not require additional formatting changes).

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/CopyingSniff.php:
```diff
@@ -0,0 +1,81 @@
+
+class ZendCodingStandard_Sniffs_Files_CopyingSniff implements \PHP_CodeSniffer_Sniff
+{
+    private $copyrightFile;
```
Needs a `@var` annotation.

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/LicenseSniff.php:
```diff
@@ -0,0 +1,81 @@
+
+class ZendCodingStandard_Sniffs_Files_LicenseSniff implements \PHP_CodeSniffer_Sniff
+{
+    private $licenseFile;
```
Needs a `@var` annotation.

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/LicenseSniff.php:
```diff
@@ -0,0 +1,81 @@
+     *                  pointer is reached. Return (count($tokens) + 1) to skip
+     *                  the rest of the file.
+     */
+    public function process(\PHP_CodeSniffer_File $phpcsFile, $stackPtr)
```
Import this class.

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/CopyingSniff.php:
```diff
@@ -0,0 +1,81 @@
+     *                  pointer is reached. Return (count($tokens) + 1) to skip
+     *                  the rest of the file.
+     */
+    public function process(\PHP_CodeSniffer_File $phpcsFile, $stackPtr)
```
Import this class.

weierophinney commented on src/ZendCodingStandard/Sniffs/Commenting/FileLevelDocBlockSniff.php:
```diff
@@ -0,0 +1,295 @@
+
+use Zend\CodingStandard\Utils\LicenseUtils;
+
+class ZendCodingStandard_Sniffs_Commenting_FileLevelDocBlockSniff implements \PHP_CodeSniffer_Sniff
```
Import this class

weierophinney commented on src/ZendCodingStandard/Utils/LicenseUtils.php:
```diff
@@ -0,0 +1,145 @@
+
+use SplFileInfo;
+
+class LicenseUtils
```
All properties and methods in this class need annotations.

weierophinney commented on test/SniffTestCase.php:
```diff
@@ -0,0 +1,150 @@
+{
+    public function processAsset($asset)
+    {
+        $phpcs = new \PHP_CodeSniffer();
```
Import this class.

weierophinney commented on test/SniffTestCase.php:
```diff
@@ -0,0 +1,150 @@
+            'showSources' => true,
+        ]);
+
+        $reflection = new \ReflectionProperty($phpcs->cli, 'values');
```
Import this class.

weierophinney commented on test/SniffTestCase.php:
```diff
@@ -0,0 +1,150 @@
+        return $phpcs->processFile($asset);
+    }
+
+    public function assertErrorCount($expectedCount, \PHP_CodeSniffer_File $file)
```
Import this class

weierophinney commented on test/SniffTestCase.php:
```diff
@@ -0,0 +1,150 @@
+
+    public function assertErrorCount($expectedCount, \PHP_CodeSniffer_File $file)
+    {
+        if (! \is_int($expectedCount)) {
```
Why is this globally qualified?

I actually have no issues with globally qualifying functions and constants. However, choose one or the other, and stick with it throughout. This one jumped at me because I haven't seen it done anywhere else.

xtreamwayz replied:
> This one jumped at me because I haven't seen it done anywhere else.

Me neither. I copied a basic function from phpunit as an example and didn't notice this.

weierophinney commented on test/Sniffs/Commenting/FileLevelDocBlockSniffTest.php:
```diff
@@ -0,0 +1,110 @@
+    public function assetsProvider()
+    {
+        return [
+            'FileLevelDocBlockPass'                   => [
```
Don't pad the `=>` declaration here or with siblings to this item. Since it's a multi-line array declaration, alignment with other items of the same nesting level isn't necessary, and actually distracts.

xtreamwayz commented:
@weierophinney There is an inconsistency in the copyright line in the license files throughout the components:

```
Copyright (c) 2015-2016, Zend Technologies USA, Inc. All rights reserved.
```

Vs

```
Copyright (c) 2005-2016, Zend Technologies USA, Inc.

All rights reserved.
```

xtreamwayz commented:
@weierophinney I've ironed out the issues. The dates in COPYING and LICENSE are now updated correctly. And the text in those files are now forced to be the same as the default so we get consistency between all components. So far I've tested zend-expressive and zend-code. In 14 seconds it updated all source files and created COPYING.md.

I've ran into another consistency issue.

Expressive:
```
/**
 * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
 * @copyright https://github.com/zendframework/zend-expressive/blob/master/COPYING.md Copyright
 * @license   https://github.com/zendframework/zend-expressive/blob/master/LICENSE.md New BSD License
 */
```

zend-code:
```
/**
 * Zend Framework (http://framework.zend.com/)
 *
 * @link      http://github.com/zendframework/zf2 for the canonical source repository
 * @copyright Copyright (c) 2005-2016 Zend Technologies USA Inc. (http://www.zend.com)
 * @license   http://framework.zend.com/license/new-bsd New BSD License
 */
```
Specifically the extra text and `@see` Vs `@link`. This can be fixed as well, but I need to know what the preferred format should be.

xtreamwayz commented:
Discussion left:
- Is the text for COPYING.md correct?
- Is the text for LICENSE.md correct?
- How to include the ruleset in components?
  1. `<rule ref="vendor/zendframework/zend-coding-standard/ruleset.xml"/>`
  2. `<rule ref="vendor/zendframework/zend-coding-standard/src/ZendCodingStandard/ruleset.xml"/>`
  The first is like a wrapper and loads the second one anyway.

xtreamwayz commented on src/ZendCodingStandard/Utils/LicenseUtils.php:
```diff
@@ -0,0 +1,189 @@
+
+All rights reserved.
+
+EOF;
```
Is this text for COPYING.md correct?

weierophinney replied:
I think this can actually be truncated to only the first line. We do not reserve all rights; the `LICENSE.md` grants the user/consumer a bunch of rights, _so long as the copyright is maintained and propagated_.

xtreamwayz commented on src/ZendCodingStandard/Utils/LicenseUtils.php:
```diff
@@ -0,0 +1,189 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+EOF;
```
Is this text for LICENSE.md correct?

weierophinney replied:
`LICENSE.md` is correct.

xtreamwayz commented on ruleset.xml:
```diff
@@ -1,23 +1,5 @@
-            <property name="ignoreBlankLines" value="false"/>
-        </properties>
-    </rule>
+    <rule ref="vendor/zendframework/zend-coding-standard/src/ZendCodingStandard/ruleset.xml"/>
```
How to properly include the ruleset in other projects?

1. `<rule ref="vendor/zendframework/zend-coding-standard/ruleset.xml"/>`
2. `<rule ref="vendor/zendframework/zend-coding-standard/src/ZendCodingStandard/ruleset.xml"/>`

The official ruleset.xml file has to be located inside `src/ZendCodingStandard` otherwise PHP_CodeSniffer doesn't detect the ruleset. 
Option 1 is basically a clean wrapper to load option 2. If in the future PHP_CodeSniffer changes something we can just adjust the wrapper without changing all packages. But I don't see this happening any time soon. Option 2 loads the ruleset directly.

xtreamwayz replied:
> Have you tested another repo using this yet?

See the updated issue (first post). Tested so far on: zend-expressive, zend-code, zend-expressive-tooling, zend-mvc, zend-http

> what happens when a user decides to use a directory other than vendor

This is custom crafted for zend framework components. If someone manages to change the vendor location he needs to change composer.json the component.

Thinking about this, someone might want to use the coding standard for his own project. I could easily add a namespace check so the license will only trigger for `zendframework/<end-<component>`.

> my main concern is repositories already using it.

I haven't tested that. But I think the first travis build or local trigger of phpcs will fail. In that case running phpcbf and a new commit with the changes will fix the build again.

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/CopyingSniff.php:
```diff
@@ -0,0 +1,92 @@
+<?php
```
I'm assuming this file and class should change to `CopyrightSniff` with the suggested `COPYRIGHT.md` change.

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/CopyingSniff.php:
```diff
@@ -0,0 +1,92 @@
+ * COPYING.md Sniff
+ *
+ * - Checks and creates COPYING.md in the project root dir
+ * - Checks and fixes copyright in COPYING.md; it should be the current year
```
`s/COPYING/COPYRIGHT/g` in this docblock.

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/CopyingSniff.php:
```diff
@@ -0,0 +1,92 @@
+ * - Checks and creates COPYING.md in the project root dir
+ * - Checks and fixes copyright in COPYING.md; it should be the current year
+ */
+class ZendCodingStandard_Sniffs_Files_CopyingSniff implements Sniff
```
I'm assuming phpcs requires this naming standard?

xtreamwayz replied:
Yes. In version 3 they will use namespaces. Not sure when it will be released.

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/CopyingSniff.php:
```diff
@@ -0,0 +1,92 @@
+     *             called again on the current file until the returned stack
+     *             pointer is reached.
+     */
+    public function process(\PHP_CodeSniffer_File $phpcsFile, $stackPtr)
```
Shouldn't the typehint be `File` here? (You alias that to this class in the import statements...)

weierophinney commented on src/ZendCodingStandard/Sniffs/Files/CopyingSniff.php:
```diff
@@ -0,0 +1,92 @@
+        }
+
+        if (! $this->copyrightFile->getRealPath()) {
+            $error = 'Missing COPYING.md file in the component root dir';
```
`s/COPYING/COPYRIGHT/`

weierophinney commented on src/ZendCodingStandard/Utils/LicenseUtils.php:
```diff
@@ -0,0 +1,189 @@
+     */
+    public static function getCopyrightFile()
+    {
+        return new SplFileInfo('COPYING.md');
```
`s/COPYING/COPYRIGHT/`

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,33 @@
+    <!-- display progress -->
+    <arg value="p"/>
+    <arg name="colors"/>
+    <file>Sniffs/Files/COPYING.md</file>
```
`s/COPYING/COPYRIGHT/`

weierophinney commented on src/ZendCodingStandard/ruleset.xml:
```diff
@@ -0,0 +1,33 @@
+    </rule>
+
+    <!-- ZendCodingStandard rules -->
+    <rule ref="ZendCodingStandard.Files.Copying"/>
```
`s/COPYING/COPYRIGHT/`

weierophinney commented on test/Sniffs/Commenting/TestAsset/FileLevelDocBlockDeprecatedLinkTag.fixed.php:
```diff
@@ -0,0 +1,12 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-coding-standard for the canonical source repository
+ * @copyright https://github.com/zendframework/zend-coding-standard/blob/master/COPYING.md Copyright
```
`s/COPYING/COPYRIGHT/` and so on. I think you can likely grep across the code base for these and fix them.

xtreamwayz commented on src/ZendCodingStandard/Sniffs/Commenting/FileLevelDocBlockSniff.php:
```diff
@@ -0,0 +1,261 @@
+    public function __construct()
+    {
+        // Get current repo name from composer.json
+        $content = file_get_contents('composer.json');
```
If getpwd() returns the directory from where phpcs/phpcbf is executed than that is even better.

xtreamwayz commented:
- Updated to php_codesniffer 3
- Simplified detection of LICENSE.md and COPYRIGHT.md (no dummy files needed anymore)
- LICENSE.md and COPYRIGHT.md are now optional.

Example `phpcs.xml` config file in a component:

```xml
<?xml version="1.0"?>
<ruleset name="Zend Framework Coding Standard">
    <!-- Load ruleset -->
    <rule ref="./vendor/zendframework/zend-coding-standard/src/ZendCodingStandard/ruleset.xml"/>

    <!-- Check copyright and license files -->
    <file>./COPYRIGHT.md</file>
    <file>./LICENSE.md</file>
    <!-- Paths to check -->
    <file>./src</file>
    <file>./test</file>

    <!-- Exclude test assets -->
    <exclude-pattern>*/TestAsset/*</exclude-pattern>
</ruleset>
```

/cc @weierophinney

xtreamwayz commented:
Superseded by #4
